README
# Sql::Textify

## NAME

Sql::Textify - Runs SQL queries and get the result in text format (markdown, html)

## VERSION

Version 0.01

## SYNOPSIS

    use Sql::Textify 'textify';
    my $text = textify( $sql );

    use Sql::Textify 'textify';
    my $text = textify( $sql, {
        conn => 'dbi:connection:string',
        username => 'username',
        password => 'password',
        format => 'markdown',
    } );

    use Sql::Textify;
    my $t = Sql::Textify->new;
    my $text = $t->textify( $sql );

    use Sql::Textify;
    my $t = Sql::Textify->new(
        conn => 'dbi:connection:string',
        username => 'username',
            password => 'password',
        format => 'markdown',
    );
    my $text = $t->textify( $sql );

## SYNTAX

This module executes SQL queries and produces text output (markdown,
html). Connection details, username and password can be specified in a
C-style multiline comment inside the SQL query:

    /*
        conn="dbi:SQLite:dbname=test.sqlite3"
        username="myusername"
        password="mypassword"
    */
    select * fom gardens;

or they can be specified using the constructor:

    my $t = new Sql::Textify(
        conn => 'dbi:SQLite:dbname=test.sqlite3',
        username => 'myusername',
        password => 'mypassword'
    );
    my $text = $t->textify('select * from gardens');

multiple queries can be separated by ; and also
insert/update/create/etc. queries are supported. If the query doesn't
return any row the string 0 rows will be returned.

## OPTIONS

Sql::Textify supports a number of options to its processor which
control the behaviour of the output document.

The options for the processor are:

### format

markdown (default), html

### layout

table (default), record

### conn

Specify the DBI connection string.

### username

Specify the database username.

### password

Specify the database password.

### maxwidth

Set a maximum width for the columns when in markdown format mode. If
any column contains a string longer than maxwidth it will be cropped.

## METHODS

### new

Sql::Textify constructor, see OPTIONS sections for more information.

### textify

The main function as far as the outside world is concerned. See the
SYNTAX for details on use.

## AUTHOR

Federico, Thiella, <fthiella at gmail.com>

## BUGS

Please report any bugs or feature requests to bug-sql-textify at
rt.cpan.org, or through the web interface at
http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Sql-Textify. I will be
notified, and then you'll automatically be notified of progress on your
bug as I make changes.

## SUPPORT

You can find documentation for this module with the perldoc command.

perldoc Sql::Textify

You can also look for information at:

* RT: CPAN's request tracker (report bugs here)

http://rt.cpan.org/NoAuth/Bugs.html?Dist=Sql-Textify

* AnnoCPAN: Annotated CPAN documentation

http://annocpan.org/dist/Sql-Textify

* CPAN Ratings

http://cpanratings.perl.org/d/Sql-Textify

* Search CPAN

http://search.cpan.org/dist/Sql-Textify/

## ACKNOWLEDGEMENTS

## LICENSE AND COPYRIGHT

Copyright 2017 Federico, Thiella.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

http://www.perlfoundation.org/artistic_license_2_0

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by
the Package. If you institute patent litigation (including a
cross-claim or counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement, then
this Artistic License to you shall terminate on the date that such
litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Sql-Textify

0.04	13 September 2017
		- little fixes to make CPAN tests happy
		- new test suites, I'm not using test::most anymore
		- test sqlite only if DBD::Sqlite is installed

0.03	13 September 2017
		- better regexp for removing C-style comments
		- also there's no need for sqltextify.pl to split the string (the textify script will do everything already) so I removed the split code.

0.02	13 September 2017
		Little internal changes, updated Makefile.PL

0.01    10 September 2017
        First version, released on an unsuspecting world.
        It contains the module Sql-Textify and the script sqltextify.pl.


lib/Sql/Textify/t/Textify.pm
package Sql::Textify::t::Textify;

use Test::Class::Most parent => 'Sql::Textify::Test::Class';

sub test_basic_class_format : Tests {
	my $self = shift;

	my $t = new Sql::Textify;

    ok( defined $t,               'new() returned something' );
    ok( $t->isa('Sql::Textify'),  "and it's the right class" );
    is( $t->{format}, 'markdown', 'default format markdown'  );

    # test formats

    my @formats = (
      	{ format => 'markdown', valid => 1 },
    	{ format => 'html',     valid => 1 },
    	{ format => 'json',     valid => 0 }
    );

    foreach my $f (@formats) {

    	if ($f->{valid}) {
	    	my $t = new Sql::Textify(format => $f->{format});
    		is( $t->{format}, $f->{format}, "format $f->{format}");
    		} else {
    			dies_ok {
    				new Sql::Textify(format => $f->{format});
    			} 'dies on invalid format';
    		}
    }

}

sub test_sql_settings : Tests {
	my $self = shift;
	my $t = new Sql::Textify;

	my $sql = $self->content_for_tests('simple');
	$t->textify($sql);
	like( $t->{conn},   qr|dbi:SQLite:dbname=(.*?)/samples\.db|, 'got correct connection string');
	is( $t->{format},   'markdown', 'still got markdown format');
	is( $t->{username}, 'username', 'got correct username');
	is( $t->{password}, 'password', 'got correct password');
	is( $t->{maxwidth}, undef,      'max width correctly undefined');
}

sub test_queries : Tests {
    my $self = shift;
    my $t = new Sql::Textify;

    my $tests = $self->records_for_tests;

    foreach my $test (@{ $tests }) {

        foreach my $result (@{ $test->{results} }) {
            $t->{format} = $result->{format}[0];
            $t->{layout} = $result->{format}[1];

            is( $t->_Do_Format($test->{data}), $result->{text}, "Test name=$test->{name}, format=$result->{format}[0], layout=$result->{format}[1]");
        }
    }
}

1;


lib/Sql/Textify/Test/Class.pm
package Sql::Textify::Test::Class;

use Test::Class::Most;
use strict;
use warnings;
use Sql::Textify;
use File::Temp;

sub records_for_tests {
  my ($self) = shift;

  my $test_ref = [
  # Simple Recordset
    {
      name => 'Simple Recordset',
      data => {
        fields => ['one', 'two'],
        rows => [
          ['first', 'second']
        ],
      },
      results => [
        {
          format => ['markdown', 'table'],
          text => '
one   | two   
------|-------
first | second
',
        },
        {
          format => ['markdown', 'record'],
          text => '# Record 1

Column | Value 
-------|-------
one    | first 
two    | second

',
        },
        {
          format => ['html', 'table'],
          text => '<table>
<thead>
  <th>one</th>
  <th>two</th>
</thead>
<tbody>
<tr>
  <td>first</td>
  <td>second</td>
</tr>
</tbody>
</table>

',
        },
        {
          format => ['html', 'record'],
          text => '<h1>Record 1</h1>

<table>
<tr>
  <th>one</th>
  <td>first</td>
</tr>
<tr>
  <th>two</th>
  <td>second</td>
</tr>
</table>

',
      }
      ]
    },
  #
  # More Lines
  #
    {
      name => 'More Lines',
      data => {
        fields => ['one', 'two'],
        rows => [
          ['first', 'second'],
          ['the loneliness of the long distance runner', 'long row']
        ],
      },
      results => [
        {
          format => ['markdown', 'table'],
          text => '
one                                        | two     
-------------------------------------------|---------
first                                      | second  
the loneliness of the long distance runner | long row
',
        },
        {
          format => ['markdown', 'record'],
          text => '# Record 1

Column | Value 
-------|-------
one    | first 
two    | second

# Record 2

Column | Value                                     
-------|-------------------------------------------
one    | the loneliness of the long distance runner
two    | long row                                  

',
        },
        {
          format => ['html', 'table'],
          text => '<table>
<thead>
  <th>one</th>
  <th>two</th>
</thead>
<tbody>
<tr>
  <td>first</td>
  <td>second</td>
</tr>
<tr>
  <td>the loneliness of the long distance runner</td>
  <td>long row</td>
</tr>
</tbody>
</table>

',
        },
        {
          format => ['html', 'record'],
          text => '<h1>Record 1</h1>

<table>
<tr>
  <th>one</th>
  <td>first</td>
</tr>
<tr>
  <th>two</th>
  <td>second</td>
</tr>
</table>

<h1>Record 2</h1>

<table>
<tr>
  <th>one</th>
  <td>the loneliness of the long distance runner</td>
</tr>
<tr>
  <th>two</th>
  <td>long row</td>
</tr>
</table>

',
      }
      ]
    },

  #
  # Quote Markdown and Html
  #
    {
      name => 'Quote Markdown and Html',
      data => {
        fields => ['column\'with"quotes|spaces and pipes', 'column<html>'],
        rows => [
          ['quotes \'"|space', '<h1></h1>'],
        ],
      },
      results => [
        { # there's not a standard way to quote markdown text
          # stackedit ignores html tags but interpreters correctly the quoted pipes .. markdown here ignores everyting
          format => ['markdown', 'table'],
          text => '
column\'with"quotes\|spaces and pipes | column<html>
-------------------------------------|-------------
quotes \'"\|space                     | <h1></h1>   
',
        },
        {
          format => ['html', 'table'],
          text => '<table>
<thead>
  <th>column&#39;with&quot;quotes|spaces and pipes</th>
  <th>column&lt;html&gt;</th>
</thead>
<tbody>
<tr>
  <td>quotes &#39;&quot;|space</td>
  <td>&lt;h1&gt;&lt;/h1&gt;</td>
</tr>
</tbody>
</table>

',
        },
      ]
    },
  ];

  return $test_ref;
}

sub content_for_tests {
    my ($self, $want) = @_;

    my $simple_src = <<SIMPLE;
/*
  conn="dbi:SQLite:dbname=tmp/samples.db"
  username="username"
  password="password"
*/
select 'first' as one, 'second' as two
SIMPLE

    my $simple_markdown = <<SIMPLE_MARKDOWN;

one   | two   
------|-------
first | second
SIMPLE_MARKDOWN

    my $simple_markdown_record = <<SIMPLE_MARKDOWN_RECORD;
# Record 1

Column | Value 
-------|-------
one    | first 
two    | second

SIMPLE_MARKDOWN_RECORD

    my $simple_html = <<SIMPLE_HTML;
<table>
<thead>
  <th>one</th>
  <th>two</th>
</thead>
<tbody>
<tr>
  <td>first</td>
  <td>second</td>
</tr>
</tbody>
</table>

SIMPLE_HTML

    my $simple_html_record = <<SIMPLE_HTML_RECORD;
<h1>Record 1</h1>

<table>
<tr>
  <th>one</th>
  <td>first</td>
</tr>
<tr>
  <th>two</th>
  <td>second</td>
</tr>
</table>

SIMPLE_HTML_RECORD

    my $max_src = <<SETWIDTH;
/*
  conn="dbi:SQLite:dbname=tmp/samples.db"
  username="username"
  password="password"
  maxwidth="1"
*/
select 'first' as one, 'second' as two
SETWIDTH

    my $max_markdown = <<SETWIDTH_MARKDOWN;

o | t
--|--
f | s
SETWIDTH_MARKDOWN

    my $max_markdown_record = <<SETWIDTH_MARKDOWN_RECORD;
# Record 1

C | V
--|--
o | f
t | s

SETWIDTH_MARKDOWN_RECORD

    my $create_src = <<CREATE;
/*
  conn="dbi:SQLite:dbname=tmp/samples.db"
  username="username"
  password="password"
  maxwidth="1"
*/
drop table if exists test;

create table test (
  id int,
  description text
);

insert into test values
(1, 'first row'),
(2, 'second row'),
(3, 'third row');
CREATE


    my $create_markdown = <<CREATE_MARKDOWN;
0 rows
0 rows
0 rows
CREATE_MARKDOWN

	my $dir = File::Temp::tempdir( CLEANUP => 1 );

	$simple_src =~ s/\btmp\b/$dir/;
	$max_src =~ s/\btmp\b/$dir/;
	$create_src =~ s/\btmp\b/$dir/;

    return $simple_src               if $want eq 'simple';
    return $simple_markdown          if $want eq 'simple-markdown';
    return $simple_markdown_record   if $want eq 'simple-markdown-record';
    return $simple_html              if $want eq 'simple-html';
    return $simple_html_record       if $want eq 'simple-html-record';
    return $max_src                  if $want eq 'max';
    return $max_markdown             if $want eq 'max-markdown';
    return $max_markdown_record      if $want eq 'max-markdown-record';
    return $create_src               if $want eq 'create';
    return $create_markdown          if $want eq 'create-markdown';
    die "No content for '$want'";
}

1;

Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Sql::Textify',
    AUTHOR           => q{Federico, Thiella <fthiella@gmail.com>},
    VERSION_FROM     => 'lib/Sql/Textify.pm',
    ABSTRACT_FROM    => 'lib/Sql/Textify.pm',
    LICENSE          => 'artistic_2',
    PL_FILES         => {},
    EXE_FILES        => ['script/sqltextify.pl'],
    MIN_PERL_VERSION => '5.006',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => '0',
    },
    BUILD_REQUIRES => {
        'Test::More' => '0',
    },
#    TEST_REQUIRES => {
#        'Test::Class::Most' => '0',
#    },
    PREREQ_PM => {
        'DBI' => '0',
        'HTML::Entities' => '0',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Sql-Textify-*' },
);


script/sqltextify.pl
#!/usr/bin/perl

=head1 NAME

sqltextify.pl - Shell interface to SQL::Textify, run a SQL query and get the result in text format (markdown, html, csv)

=head1 SYNOPSIS

    sqltextify.pl --sql=query.sql --conn="dbi:SQLite:dbname=test.sqlite3" --username=admin --password=pass

=head1 DESCRIPTION

sqltextify.pl is a shell interface to SQL::Textify, it will run one or more SQL queries and will return the result
in various text formats (Markdown and HTML at the moment, both using a table layout or a record layout).
This can be used at the command line or can be easily integrated with Sublime Text.

=head1 LICENSE

This is released under the Artistic 
License. See L<perlartistic>.

=head1 AUTHOR

Federico Thiella - GitHub projects L<https://github.com/fthiella/>
or email L<mailto:fthiella@gmail.com>

=cut

use strict;
use warnings;
 
use DBI;
use File::Slurp;
use Getopt::Long;
use utf8;
use Sql::Textify;

our $VERSION = "0.04";
our $RELEASEDATE = "September 13th, 2017";

=head1 OPTIONS

sqltextify.pl supports a number of command line options which control
the behaviour of the output document.

The options are:

=over

=item source, s

source SQL file

=item conn, c

DBI connection string

=item username, u

Specify the database username.

=item password, p

Specify the database password.

=item maxwidth, mw

Set a maximum width for the columns when in markdown format mode. If any column contains
a string longer than maxwidth it will be cropped.

=item format, f

Output format (markdown default, html)

=item layout, l

Output layout (table default, record)

=back
=cut

# CLI Interface

sub do_help {
	print <<endhelp;
Usage: sqlbuild.pl [options]
       perl sqlbuild.pl [options]

Options:
  -s, --sql         source SQL file
  -c, --conn        specify DBI connection string
  -u, --username    specify username
  -p, --password    specify password
  -mw, --maxwidth   maximum width column for markdown (if unspecified get from actual data)
  -f, --format      output format (markdown -default- or html)
  -l, --layout      output layout (table -default- or record)
  -v, --version     show current version

Project GitHub page: https://github.com/fthiella/perl-Sql-Textify
endhelp
}

sub do_version {
	print "sqltextify $VERSION ($RELEASEDATE)\n";
}

# added utf8 support
# FIXME: still need to verify if it's always working
use open ':std', ':encoding(UTF-8)';

# Get command line options
my $source;
my $version;
my $conn;
my $username;
my $password;
my $maxwidth;
my $format;
my $layout;
my $help;

GetOptions(
	'sql|s=s'      => \$source,
	'version|v'    => \$version,
	'conn|c=s'     => \$conn,
	'username|u=s' => \$username,
	'password|p=s' => \$password,
	'maxwidth|w=i' => \$maxwidth,
	'format|f=s'   => \$format,
	'layout|l=s'   => \$layout,
	'help|h'       => \$help,
);

if ($help)
{
	do_help;
	exit;
}

if ($version)
{
	do_version;
	exit;
}

die "Please specfy sql source with -s or -sql\n" unless ($source);

# read the input file
my $sql = read_file($source);

# FIXME: over simplified regex, and also need to decide which parameter has to take precedence, the command line or the C-comment?

unless ($conn)     { ($conn) = $sql =~ /conn=\"([^\""]*)\"\s/; }
unless ($username) { ($username) = $sql =~ /username=\"([^\""]*)\"\s/; }
unless ($password) { ($password) = $sql =~ /password=\"([^\""]*)\"\s/; }
unless ($maxwidth) { ($maxwidth) = $sql =~ /maxwidth=\"([^\""]*)\"\s/; }
unless ($format)   { ($format) = $sql =~ /format=\"([^\""]*)\"\s/; }
unless ($layout)   { ($layout) = $sql =~ /layout=\"([^\""]*)\"\s/; }

# default
unless ($format)   { $format = 'markdown'; }
unless ($layout)   { $layout = 'table';   }

my $dbh = DBI->connect($conn, $username, $password)
|| die $DBI::errstr;


my $t = new Sql::Textify(
	conn => $conn,
	username => $username,
	password => $password,
	maxwidth => $maxwidth,
	format => $format,
	layout => $layout
);

print $t->textify($sql), "\n";


lib/Sql/Textify.pm
package Sql::Textify;

use 5.006;
use strict;
use warnings;
use Carp qw(croak);
use DBI;
use HTML::Entities;

=head1 NAME

Sql::Textify - Run SQL queries and get the result in text format (markdown, html)

=head1 VERSION

Version 0.04

=cut

our $VERSION = '0.04';
$VERSION = eval $VERSION;
our @EXPORT_OK = qw(textify);

=head1 SYNOPSIS

    use Sql::Textify;
    my $t = Sql::Textify->new;
    my $text = $t->textify( $sql );

    use Sql::Textify;
    my $t = Sql::Textify->new(
        conn => 'dbi:connection:string',
        username => 'username',
	    password => 'password',
        format => 'markdown',
    );
    my $text = $t->textify( $sql );

=head1 SYNTAX

This module executes SQL queries and produces text output (markdown, html).
Connection details, username and password can be specified in a C-style multiline
comment inside the SQL query:

    /*
        conn="dbi:SQLite:dbname=test.sqlite3"
        username="myusername"
        password="mypassword"
    */
    select * fom gardens;

or they can be specified using the constructor:

    my $t = new Sql::Textify(
        conn => 'dbi:SQLite:dbname=test.sqlite3',
        username => 'myusername',
        password => 'mypassword'
    );
    my $text = $t->textify('select * from gardens');

multiple queries can be separated by C<< ; >> and also insert/update/create/etc. queries are
supported. If the query doesn't return any row the string C<< 0 rows >> will be returned.

=head1 OPTIONS

Sql::Textify supports a number of options to its processor which control
the behaviour of the output document.

The options for the processor are:

=over

=item format

markdown (default), html

=item layout

table (default), record

=item conn

Specify the DBI connection string.

=item username

Specify the database username.

=item password

Specify the database password.

=item maxwidth

Set a maximum width for the columns when in markdown format mode. If any column contains
a string longer than maxwidth it will be cropped.

=back

=head1 METHODS

=head2 new

Sql::Textify constructor, see OPTIONS sections for more information.

=cut

sub new {
	my ($class, %p) = @_;
	my $self = {};

	$self->{format} = sanitize_string({ value => $p{format}, regexp => 'html|markdown', default => 'markdown' });
	$self->{layout} = sanitize_string({ value => $p{layout}, regexp => 'table|record',  default => 'table' });

	$self->{username} = $p{username};
	$self->{password} = $p{password};
	$self->{conn} = $p{conn};

	bless $self, ref($class) || $class;
	return $self;
}

=head2 textify

The main function as far as the outside world is concerned. See the SYNTAX
for details on use.

=cut

sub textify {
	my ( $self, $sql ) = @_;

	$self->_GetParametersFromSql($sql);
	$self->{dbh} = DBI->connect($self->{conn}, $self->{username}, $self->{password}) || die $DBI::errstr;

	return $self->_Textify($sql);
}

sub _GetParametersFromSql {
	my ($self, $sql) = @_;

	# FIXME: values from SQL string will take precedence
	# FIXME: the following regexps will usually work on most cases, but are over-simplified

	if ($sql =~ /conn=\"([^\""]*)\"\s/)     { $self->{conn} = $1;     }
	if ($sql =~ /username=\"([^\""]*)\"\s/) { $self->{username} = $1; }
	if ($sql =~ /password=\"([^\""]*)\"\s/) { $self->{password} = $1; }
	if ($sql =~ /maxwidth=\"([^\""]*)\"\s/) { $self->{maxwidth} = $1; }
	if ($sql =~ /format=\"([^\""]*)\"\s/)   { $self->{format} = $1;   }
	if ($sql =~ /layout=\"([^\""]*)\"\s/)   { $self->{layout} = $1;   }
}

sub _Do_Sql {
	my ($self, $sql_query) = @_;
	my %r;

	my $qry = $self->{dbh}->prepare($sql_query)	|| die "````\n", $sql_query, "\n````\n\n", ">", $DBI::errstr, "\n";
	$qry->execute()	|| die "````\n", $sql_query, "\n````\n\n", ">", $DBI::errstr;
	my $rows = $qry->fetchall_arrayref();
	$qry->finish();

	$r{fields} = $qry->{NAME};
	$r{rows} = $rows;

	return wantarray ? %r : \%r;
}

sub _Textify {
	my ($self, $sql) = @_;

	my $result;

	# strip C-style comments from source query
	# regexp from http://learn.perl.org/faq/perlfaq6.html#How-do-I-use-a-regular-expression-to-strip-C-style-comments-from-a-file

	$sql =~ s#/\*[^*]*\*+([^/*][^*]*\*+)*/|("(\\.|[^"\\])*"|'(\\.|[^'\\])*'|.[^/"'\\]*)#defined $2 ? $2 : ""#gse;

	foreach my $sql_query (split /;\s*/, $sql) {
		my $records = $self->_Do_Sql($sql_query);
		$result .= $self->_Do_Format($records);
 	}

	return $result;
}

sub _Do_Format {
	my ($self, $records) = @_;

	my %m = (
		"markdown" => {
			"table"  => sub { _Do_Sql_Markdown(@_) },
			"record" => sub { _Do_Sql_Markdown_Record(@_) }
		},
		"html" => {
			"table"  => sub { _Do_Sql_Html(@_) },
			"record" => sub { _Do_Sql_Html_Record(@_) }
		},
	);

	if ($m{ $self->{format} }->{ $self->{layout} }) {
		return $m{ $self->{format} }->{ $self->{layout} }->($self, $records);
	} else {
		die("Wrong format/layout (".$self->{format}."/".$self->{layout});
	}
}

sub _Do_Sql_Markdown {
	my ($self, $records) = @_;

	my $max_width = $self->{maxwidth};
	my $max_format = '';

	my $result = '';

	if (($max_width) && ($max_width> 0))
	{
		$max_format = ".$max_width";
	}

	my @width = map { min(length(quote_markdown($_)), $max_width) } @{$records->{fields}};

	if (scalar @{ $records->{rows} }>0) {

		foreach my $row ( @{ $records->{rows} }) {
			# <- for each row, and for each column -> calculate the maximum width
			foreach my $i (0 .. (scalar @{ $row }-1)) {
				$width[$i] = max($width[$i], length(quote_markdown($row->[$i])));
			}
		}
		# then "strip" the width to the $max_width
		@width = map { min($_, $max_width) } @width;

		# create format string
		my $f = join(' | ', map { "%-".$_.$max_format."s"} @width) . "\n";

		# print header
		$result .= "\n" . sprintf( $f, map { quote_markdown($_) } @{$records->{fields}} );

		# print sub header -|-
		$result .= join("-|-", map { '-'x$_ } @width ) . "\n";

		# print rows
		foreach my $row (@{ $records->{rows} }) {
			$result .= sprintf( $f, map { quote_markdown($_) } @{$row} );
		}
	} else {
		$result .= "0 rows\n";
	}

	return $result;
}

sub _Do_Sql_Markdown_Record {
	my ($self, $records) = @_;

	my $max_width = $self->{maxwidth};

	my $max_format = '';
	my $nr = 1;

	my $result = '';

	if (($max_width) && ($max_width > 0))
	{
		$max_format = ".$max_width";
	}

	# rows
	foreach my $row (@{ $records->{rows} }) {
		$result .= "# Record $nr\n\n";

		my @width = (min(length('Column'),$max_width),min(length('Value'),$max_width));

		foreach my $i (0 .. (scalar @{ $row }-1)) {
			$width[0]=max($width[0], length(quote_markdown($records->{fields}[$i])));
			$width[1]=max($width[1], length(quote_markdown($row->[$i])),$max_width);
		}
		$width[0] = min($width[0], $max_width);
		$width[1] = min($width[1], $max_width);

		# %-x.ys    %s string, - left-justify, x minimum widht, y maximum width
		my $f = join(' | ', map { "%-".$_.$max_format."s"} @width) . "\n";

		$result .= sprintf $f, ("Column", "Value");

		$result .= '-'x$width[0] . '-|-' . '-'x$width[1] . "\n";

		foreach my $i (0 .. (scalar @{ $row }-1)) {
				$result .= sprintf $f, (quote_markdown($records->{fields}[$i]), quote_markdown($row->[$i]));
		}

		$result .= "\n";
		$nr++;
	}

	if ($nr == 1) { # no rows returned, empty rowset or create/update query
		$result .= "0 rows\n";
	}

	return $result;
}

sub _Do_Sql_Html {
	my ($self, $records) = @_;
	my $result;

	if (scalar @{ $records->{rows} }>0) {

		$result  = "<table>\n";
		$result .= "<thead>\n";

		$result .= join("\n", map { '  <th>' . encode_entities($_) . '</th>'} @{$records->{fields}}) . "\n";

		$result .= "</thead>\n";
		$result .= "<tbody>\n";

		foreach my $row (@{ $records->{rows} }) {
			$result .= "<tr>\n";
			$result .= join("\n", map { '  <td>' . encode_entities($_) . '</td>'} @{ $row }) . "\n";
			$result .= "</tr>\n";
		}

		$result .= "</tbody>\n";
		$result .= "</table>\n\n";
	} else {
		$result .= "<p>\n0 rows</p>\n\n";
	}

	return $result;
}

sub _Do_Sql_Html_Record {
	my ($self, $records) = @_;
	my $result;
	my $nr = 1;

	if (scalar @{ $records->{rows} }>0) {
		foreach my $row (@{ $records->{rows} }) {
			$result .= "<h1>Record $nr</h1>\n\n";

			$result .= "<table>\n";

			foreach my $i (0 .. (scalar @{ $row }-1)) {
				$result .= "<tr>\n";
				$result .= "  <th>" . encode_entities($records->{fields}[$i]) . "</th>\n";
				$result .= "  <td>" . encode_entities($row->[$i]) . "</td>\n";
				$result .= "</tr>\n";
			}

			$result .= "</table>\n\n";

			$nr++;
		}
	} else {
		$result .= "<p>\n0 rows</p>\n\n";
	}

	return $result;
}

# internal functions

sub sanitize_string {
    my $p = shift;

	return $p->{default} unless $p->{value};
	return $p->{value}   if $p->{value} =~ /^$p->{regexp}$/;
	croak("Invalid value *$p->{value}* provided.");
}

sub quote_markdown {
	# there's not a standard way to quote markdown
	my $s = shift;

	if ($s) {
		# quote |
		$s =~ s/\|/\\\|/g;

		# replace non-printable characters with space
		$s =~ s/[^[:print:]]/ /g;
	} else {
		$s = '';
	}
	return $s;
}

sub max ($$) {
	# if second parameter is defined then return max(p1, p2) otherwise return p1
	if ($_[1]) {
		$_[$_[0] < $_[1]];
	} else {
		$_[0];
	}
}

sub min ($$) {
	# if second parameter is defined then return min(p1, p2) otherwise return p1
	if ($_[1]) {
		$_[$_[0] > $_[1]];
	} else {
		$_[0];
	}
}

=head1 AUTHOR

Federico, Thiella, C<< <fthiella at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-sql-textify at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Sql-Textify>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Sql::Textify


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Sql-Textify>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Sql-Textify>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Sql-Textify>

=item * Search CPAN

L<http://search.cpan.org/dist/Sql-Textify/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2017 Federico, Thiella.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Sql::Textify
